Proposed 2.2.0 (#5037)
XRPLF/rippled: RELEASENOTES.md

Have new ideas? Need help with setting up your node? [Please open an issue here](https://github.com/xrplf/rippled/issues/new/choose).

# Version 2.2.0

> Version 2.2.0 of `rippled`, the reference server implementation of the XRP Ledger protocol, is now available. This release adds the price oracle, performance optimizations, several a bug fixes, and introduces the `featurePriceOracle`, `fixEmptyDID`, `fixXChainRewardRounding`, `fixPreviousTxnID`, and `fixAMMv1` amendments.

Reviewer: nits:
```suggestion
Version 2.2.0 of `rippled`, the reference server implementation of the XRP Ledger protocol, is now available. This release adds the price oracle, performance optimizations, several bug fixes, and introduces the `featurePriceOracle`, `fixEmptyDID`, `fixXChainRewardRounding`, `fixPreviousTxnID`, and `fixAMMv1` amendments.
```

Price Oracle is mentioned twice in this sentence here, might be redundant

Author: Yeah, I was listing out all the amendments, but I also wanted to highlight the price oracle. But I agree it doesn't read well. I took out the first reference. Fixed (force pushed).

> Version 2.2.0 of `rippled`, the reference server implementation of the XRP Ledger protocol, is now available. This release adds the price oracle, performance optimizations, several a bug fixes, and introduces the `featurePriceOracle`, `fixEmptyDID`, `fixXChainRewardRounding`, `fixPreviousTxnID`, and `fixAMMv1` amendments.

Reviewer 2: Nit: should be `fixAMMv1_1`, not `fixAMMv1`

[Sign Up for Future Release Announcements](https://groups.google.com/g/ripple-server)

<!-- BREAK -->

## Action Required

Five new amendments are now open for voting according to the XRP Ledger's [amendment process](https://xrpl.org/amendments.html), which enables protocol changes following two weeks of >80% support from trusted validators.

If you operate an XRP Ledger server, upgrade to version 2.2.0 by June 17, 2024 to ensure service continuity. The exact time that protocol changes take effect depends on the voting decisions of the decentralized network.

> If you operate a Clio server, Clio needs to be updated to 2.1.2 before updating to rippled 2.2.0. Clio will be blocked if it is not update.

Reviewer: nit: typo `update` => `updated`

Reviewer: tagging @cindyyan317 and @kuznetsss to review Clio update message

Reviewer 3: LGTM 👍

Reviewer 4: Looks good, thanks 👍

Author: Fixed the nix (forced pushed).

## Changelog

### Amendments and New Features
(These are changes which may impact or be useful to end users. For example, you may be able to update your code/workflow to take advantage of these changes.)

> - **featurePriceOracle**: This implements a price oracle as defined in the [XLS-47](https://github.com/XRPLF/XRPL-Standards/blob/master/XLS-47d-PriceOracles/README.md) spec. A Price Oracle is used to bring real-world data, such as market prices, onto the blockchain, enabling dApps to access and utilize information that resides outside the blockchain. [#4789](https://github.com/XRPLF/rippled/pull/4789) 
> 
> - **fixEmptyDID**: Adds an additional check to ensure that DIDs are non-empty when created, and returns a `tecEMPTY_DID` error if the DID would be empty. [#4950](https://github.com/XRPLF/rippled/pull/4950)
> 
> - **fixXChainRewardRounding**: Fixes rounding so reward shares are always rounded down, even when the `fixUniversalNumber` amendment is active. [#4933](https://github.com/XRPLF/rippled/pull/4933)
> 
> - **fixPreviousTxnID**: Adds `PreviousTxnID` and `PreviousTxnLgrSequence` as fields to all ledger objects that did not already have them included (`DirectoryNode`, `Amendments`, `FeeSettings`, `NegativeUNL`, and `AMM`). [#4751](https://github.com/XRPLF/rippled/pull/4751)
> 
> - **fixAMMv1_1**: Fixes AMM offer rounding and low quality LOB offers from blocking the AMM. [#4983](https://github.com/XRPLF/rippled/pull/4983)

Reviewer 5: Several suggestions on these amendment descriptions:

1. Reiterate that these are amendments here to help comprehension while skimming. Otherwise people might think the described changes go into effect immediately.
2. Add more context for the amendments that modify the behavior of other amendments that are currently in voting.
3. Clarify that fixPreviousTxnID applies "lazily".
4. Use "ledger entry" instead of "ledger object".
5. Spell out "order book" instead of using jargon ("LOB")

```suggestion
- **featurePriceOracle** amendment: Implements a price oracle as defined in the [XLS-47](https://github.com/XRPLF/XRPL-Standards/blob/master/XLS-47d-PriceOracles/README.md) spec. A Price Oracle is used to bring real-world data, such as market prices, onto the blockchain, enabling dApps to access and utilize information that resides outside the blockchain. [#4789](https://github.com/XRPLF/rippled/pull/4789) 

- **fixEmptyDID** amendment: Modifies the behavior of the DID amendment: adds an additional check to ensure that DIDs are non-empty when created, and returns a `tecEMPTY_DID` error if the DID would be empty. [#4950](https://github.com/XRPLF/rippled/pull/4950)

- **fixXChainRewardRounding** amendment: Modifies the behavior of the XChainBridge amendment: fixes rounding so reward shares are always rounded down, even when the `fixUniversalNumber` amendment is active. [#4933](https://github.com/XRPLF/rippled/pull/4933)

- **fixPreviousTxnID** amendment: Adds `PreviousTxnID` and `PreviousTxnLgrSequence` as fields to all ledger entries that did not already have them included (`DirectoryNode`, `Amendments`, `FeeSettings`, `NegativeUNL`, and `AMM`). Existing ledger entries will gain the fields whenever transactions modify those entries. [#4751](https://github.com/XRPLF/rippled/pull/4751). 

- **fixAMMv1_1** amendment: Fixes AMM offer rounding and low quality order book offers from blocking the AMM. [#4983](https://github.com/XRPLF/rippled/pull/4983)
```

> - Add user version of `feature` RPC. [#4781](https://github.com/XRPLF/rippled/pull/4781)

Reviewer 5: ```suggestion
- Add a non-admin version of `feature` API method. [#4781](https://github.com/XRPLF/rippled/pull/4781)
```

### Bug Fixes and Performance Improvements
(These are behind-the-scenes improvements, such as internal changes to the code, which are not expected to impact end users.)

> - Optimized the base58 encoder and decoder. This is not about an order of magnitude faster than the existing algorithm (10x faster for encoding and 15x faster for decoding) [#4327](https://github.com/XRPLF/rippled/pull/4327)

Reviewer 2: "This is not about an order of magnitude" - sounds confusing to me non-English ears.

Reviewer 6: I think this is the problem:
```suggestion
- Optimized the base58 encoder and decoder. This is about an order of magnitude faster than the existing algorithm (10x faster for encoding and 15x faster for decoding) [#4327](https://github.com/XRPLF/rippled/pull/4327)
```

> - Optimized the base58 encoder and decoder. This is now about an order of magnitude faster than the existing algorithm (10x faster for encoding and 15x faster for decoding) [#4327](https://github.com/XRPLF/rippled/pull/4327)
> 
> - Optimize the `account_tx` SQL query. [#4955](https://github.com/XRPLF/rippled/pull/4955)
> 
> - Don't reach consensus as quickly if no other proposals are seen. [#4763](https://github.com/XRPLF/rippled/pull/4763)
> 
> - Fixed a potential deadlock in the database module. [#4989](https://github.com/XRPLF/rippled/pull/4989)
> 
> - Enforce no duplicate slots from incoming connections. [#4944](https://github.com/XRPLF/rippled/pull/4944)
> 
> - Fixed an order book update variable swap. [#4890](https://github.com/XRPLF/rippled/pull/4890)

Reviewer 5: Here are some copy edits for consistency in verb tense and brevity.

```suggestion
- Optimize the base58 encoder and decoder. The algorithm is now about 10 times faster for encoding and 15 times faster for decoding. [#4327](https://github.com/XRPLF/rippled/pull/4327)

- Optimize the `account_tx` SQL query. [#4955](https://github.com/XRPLF/rippled/pull/4955)

- Don't reach consensus as quickly if no other proposals are seen. [#4763](https://github.com/XRPLF/rippled/pull/4763)

- Fix a potential deadlock in the database module. [#4989](https://github.com/XRPLF/rippled/pull/4989)

- Enforce no duplicate slots from incoming connections. [#4944](https://github.com/XRPLF/rippled/pull/4944)

- Fix an order book update variable swap. [#4890](https://github.com/XRPLF/rippled/pull/4890)
```

### Docs and Build System

> - Added unit test to raise the test coverage of the AMM. [#4971](https://github.com/XRPLF/rippled/pull/4971)
> 
> - Improved test coverage reporting. [#4977](https://github.com/XRPLF/rippled/pull/4977)

Reviewer 5: Nit: match the verb tense in the above sections.

```suggestion
- Add unit test to raise the test coverage of the AMM. [#4971](https://github.com/XRPLF/rippled/pull/4971)

- Improve test coverage reporting. [#4977](https://github.com/XRPLF/rippled/pull/4977)
```
